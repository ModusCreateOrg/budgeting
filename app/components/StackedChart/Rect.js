// @flow
import * as React from 'react';
import { select, interpolate } from 'd3';

type RectProps = {
  width: number,
  height: number,
  fill: string,
  y: number,
  animDuration: number,
};

class Rect extends React.Component<RectProps> {
  static defaultProps = {
    animDuration: 1000,
  };

  componentDidMount() {
    const { animDuration } = this.props;

    const rect = select(this.rectRef);

    if (this.rectRef) {
      const interpolateHeight = interpolate(1000, this.rectRef.getAttribute('height'));

      rect
        .transition()
        .duration(animDuration * Math.random())
        .attrTween('height', () => t => interpolateHeight(t));
    }
  }

  rectRef: ?HTMLElement;

  handleRefUpdate = (ref: ?HTMLElement) => {
    this.rectRef = ref;
  };

  render() {
    const { width, height, fill, y } = this.props;
    return <rect ref={this.handleRefUpdate} {...{ y, height, width, fill }} />;
  }
}

export default Rect;
