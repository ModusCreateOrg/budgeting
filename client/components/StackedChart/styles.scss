@import 'theme/variables';

.stackedChart {
  display: flex;
  height: 100%;
  justify-content: center;
  align-content: flex-start;
  flex-flow: wrap;
}

.xAxis {
   font-size: 16px;
   font-family: $font-default;
   font-weight: bold;
   text-anchor: middle;

  .value {
     font-size: 16px;
  }
}
