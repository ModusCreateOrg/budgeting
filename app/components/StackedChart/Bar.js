// @flow
import * as React from 'react';
import type { TransactionSummary } from 'selectors/transactions';
import Rect from './Rect';

type BarProps = {
  yScale: Function,
  colorFn: Function,
  data: TransactionSummary[],
  transform: string,
  width: number,
};

class Bar extends React.Component<BarProps> {
  componentWillMount() {
    this.updateChartVariables();
  }

  componentWillReceiveProps(newProps: BarProps) {
    const { yScale, data } = newProps;
    const old = this.props;

    if (old.yScale !== yScale || old.data !== data) {
      this.updateChartVariables();
    }
  }

  yPositions: Array<any>;

  updateChartVariables = () => {
    const { yScale, data } = this.props;
    let start = yScale.range()[0];

    this.yPositions = data.map(datum => {
      start -= yScale(datum.value);
      return start;
    });
  };

  render() {
    const { yPositions } = this;
    const { width, yScale, colorFn, data, transform } = this.props;

    return (
      <g transform={transform}>
        {data.map((datum, idx) => (
          <Rect
            key={datum.categoryId}
            y={yPositions[idx]}
            height={yScale(datum.value)}
            width={width}
            fill={colorFn(idx)}
          />
        ))}
      </g>
    );
  }
}

export default Bar;
