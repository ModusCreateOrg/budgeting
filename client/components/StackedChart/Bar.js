import React, { Component, PropTypes } from 'react';

import Rect from './Rect';

class Bar extends Component {
  static propTypes = {
    yScale: PropTypes.func.isRequired,
    colorFn: PropTypes.func.isRequired,
    data: PropTypes.array.isRequired,
    transform: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
  };

  componentWillMount() {
    this.updateChartVariables();
  }

  componentWillReceiveProps({ yScale, data }) {
    const old = this.props;

    if (old.yScale !== yScale || old.data !== data) {
      this.updateChartVariables();
    }
  }

  updateChartVariables = () => {
    const { yScale, data } = this.props;
    let start = yScale.range()[0];

    this.yPositions = data.map(datum => (start -= yScale(datum.value)));
  };

  render() {
    const { yPositions } = this;
    const { width, yScale, colorFn, data, transform } = this.props;

    return (
      <g transform={transform}>

        {data.map((datum, idx) => (
          <Rect 
            key={datum.categoryId}
            y={yPositions[idx]}
            height={yScale(datum.value)}
            width={width}
            fill={colorFn(idx)}
          />
        ))}

      </g>
    );
  }
}

export default Bar;
