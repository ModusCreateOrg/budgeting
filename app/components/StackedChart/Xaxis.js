// @flow
import * as React from 'react';
import type { TransactionSummary } from 'selectors/transactions';
import formatAmount from 'utils/formatAmount';
import styles from './styles.scss';

type XaxisProps = {
  transform: string,
  labelColor: string,
  valueColor: string,
  data: {
    inflow: TransactionSummary[],
    outflow: TransactionSummary[],
  },
  totals: Object,
  xScale: Function,
};

const Xaxis = ({ data, totals, transform, labelColor, valueColor, xScale }: XaxisProps) => (
  <g className={styles.xAxis} transform={transform}>
    {Object.keys(data).map((key, idx) => (
      <g key={key} transform={`translate(${xScale(idx) + xScale.bandwidth() / 2}, 0)`}>
        <line stroke={labelColor} y2="6" x1="0.5" x2="0.5" />
        <text fill={labelColor} y="9" x="0.5" dy="0.8em">
          {key.toUpperCase()}
        </text>
        <text className={styles.value} fill={valueColor} y="35" x="0.5" dy="0.6em">
          {formatAmount(totals[key]).text}
        </text>
      </g>
    ))}
  </g>
);

Xaxis.defaultProps = {
  labelColor: '#000',
  valueColor: '#999',
  transform: '',
};

export default Xaxis;
