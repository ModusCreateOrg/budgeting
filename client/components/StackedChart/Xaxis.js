import React, { PropTypes } from 'react';

import styles from './styles.scss';


const Xaxis = ({ className, transform, color, data, xScale }) => (
  <g {...{className, transform}}>

    {Object.keys(data).map((key, idx) => (      
      <g key={key} transform={`translate(${xScale(idx) + xScale.bandwidth() / 2}, 0)`}>

        <line stroke={color} y2="6" x1="0.5" x2="0.5" />
        <text fill={color} y="9" x="0.5" dy="0.71em">
          {key.toUpperCase()}
        </text>

      </g>
    ))}

  </g>
);


Xaxis.propTypes = {
  className: PropTypes.string,
  transform: PropTypes.string,
  color: PropTypes.string,
  data: PropTypes.object.isRequired,
  xScale: PropTypes.func.isRequired,  
};

Xaxis.defaultProps = {
  color: '#000'
}

export default Xaxis;
