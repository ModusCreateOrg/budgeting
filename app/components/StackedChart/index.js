// @flow
import * as React from 'react';
import type { TransactionSummary } from 'selectors/transactions';
import Legend from 'components/Legend';
import Chart from 'components/Chart';
import { max, scaleBand, scaleLinear, scaleOrdinal, schemeCategory20 } from 'd3';
import { shuffle } from 'utils/array';
import Bar from './Bar';
import Xaxis from './Xaxis';
import styles from './styles.scss';

const outflowScheme = shuffle([...schemeCategory20.slice(0, 4), ...schemeCategory20.slice(5)]);
const inflowScheme = ['#2ca02c']; // inflow always green

type StackedChartProps = {
  width: number,
  height: number,
  dataValue: string,
  dataLabel: string,
  dataKey: string,
  data: {
    inflow: TransactionSummary[],
    outflow: TransactionSummary[],
  },
  totals: {
    inflow: number,
    outflow: number,
  },
};

class StackedChart extends React.Component<StackedChartProps> {
  static defaultProps = {
    width: 300,
    height: 500,
    dataValue: 'value',
  };

  componentWillMount() {
    this.updateChartVariables();
  }

  componentWillReceiveProps(newProps: StackedChartProps) {
    const old = this.props;

    if (old.data !== newProps.data) {
      this.updateChartVariables();
    }
  }

  dataKeys: string[];
  xScale: Function;
  yScale: Function;
  colorFn: any;
  boxLength: number;
  boxHeight: number;

  updateChartVariables = () => {
    const { width, height, data, totals } = this.props;
    const { color, barPadding, chartPadding, bottomPadding } = this;

    this.dataKeys = Object.keys(data);

    this.xScale = scaleBand()
      .rangeRound([0, width - chartPadding * 2])
      .paddingInner(barPadding);
    this.xScale.domain([0, this.dataKeys.length - 1]);

    this.yScale = scaleLinear().rangeRound([height - chartPadding * 2 - bottomPadding, 0]);
    this.yScale.domain([max([totals.inflow, totals.outflow]), 0]);

    this.colorFn = this.dataKeys.reduce((colorFn, key) => {
      colorFn[key] = color[key].domain([0, data[key].length]);
      return colorFn;
    }, {});

    this.boxLength = width + chartPadding * 2;
    this.boxHeight = height + chartPadding * 2;
  };

  barPadding = 0.15;
  bottomPadding = 40;
  chartPadding = 10;

  color = {
    inflow: scaleOrdinal(inflowScheme),
    outflow: scaleOrdinal(outflowScheme),
  };

  render() {
    const { xScale, yScale, colorFn, dataKeys, boxLength, boxHeight, chartPadding } = this;
    const { data, totals, dataKey, dataLabel, dataValue } = this.props;

    return (
      <div className={styles.stackedChart}>
        <Chart
          width={boxLength}
          height={boxHeight}
          padding={chartPadding}
          transform={`translate(${chartPadding},${chartPadding})`}
        >
          {dataKeys.map((key, idx) => (
            <Bar
              key={key}
              data={data[key]}
              yScale={yScale}
              colorFn={colorFn[key]}
              width={xScale.bandwidth()}
              transform={`translate(${xScale(idx)}, 0)`}
            />
          ))}

          <Xaxis
            transform={`translate(0, ${yScale.range()[0] + chartPadding / 3})`}
            data={data}
            totals={totals}
            xScale={xScale}
          />
        </Chart>

        <Legend color={colorFn.outflow} reverse data={data.outflow} {...{ dataValue, dataLabel, dataKey }} />
      </div>
    );
  }
}

export default StackedChart;
