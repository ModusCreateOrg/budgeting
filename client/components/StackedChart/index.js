import React, { Component, PropTypes } from 'react';

import {
  max,
  scaleBand,
  scaleLinear,
  scaleSequential,
  interpolateMagma
} from 'd3';

import Bar from './Bar';
import Xaxis from './Xaxis';
import Legend from 'components/Legend';
import Chart from 'components/Chart'
import styles from './styles.scss';


class StackedChart extends Component {
  static propTypes = {
    width: PropTypes.number,
    height: PropTypes.number,
    dataValue: PropTypes.string,
    dataLabel: PropTypes.string.isRequired,
    dataKey: PropTypes.string.isRequired,
    data: PropTypes.shape({
      inflow: PropTypes.array,
      outflow: PropTypes.array
    }).isRequired,
  };

  static defaultProps = {
    width: 300,
    height: 500,
    dataValue: 'value'
  };

  componentWillMount() {
    this.updateChartVariables();
  }

  componentWillReceiveProps({ data }) {
    const old = this.props;

    if (old.data !== data) {
      this.updateChartVariables();
    }
  }

  barPadding = 0.15;
  chartPadding = 10;
  color = {
    inflow: scaleLinear().range(['#003300', '#009966']),
    outflow: scaleSequential().interpolator(interpolateMagma),
  }

  getTotalValue = (total, datum) => (total + datum.value);

  updateChartVariables = () => {
    const { width, height, data } = this.props;
    const { color, barPadding, chartPadding } = this;
    
    this.dataKeys = Object.keys(data);
    const totals = this.dataKeys.map(key => data[key].reduce(this.getTotalValue, 0));

    this.xScale = scaleBand().rangeRound([0, width - chartPadding * 2]).paddingInner(barPadding);
    this.xScale.domain([0, this.dataKeys.length - 1]);

    this.yScale = scaleLinear().rangeRound([height - chartPadding * 4, 0]);
    this.yScale.domain([max(totals), 0]);

    this.colorFn = this.dataKeys.reduce((colorFn, key) => {
      colorFn[key] = color[key].domain([0, data[key].length]);
      return colorFn;
    }, {});

    this.boxLength = width + (chartPadding * 2);
    this.boxHeight = height + (chartPadding * 2);
  }

  render() {
    const { xScale, yScale, colorFn, dataKeys, boxLength, boxHeight, chartPadding } = this;
    const { data, dataKey, dataLabel, dataValue } = this.props;

    return (
      <div className={styles.stackedChart}>
        <Chart 
          width={boxLength}
          height={boxHeight}
          padding={chartPadding}
          transform={`translate(${chartPadding},${chartPadding})`}
        >
          {dataKeys.map((key, idx) => (
            <Bar
              key={key}
              data={data[key]}
              yScale={yScale}
              colorFn={colorFn[key]}
              width={xScale.bandwidth()}
              transform={`translate(${xScale(idx)}, 0)`}
            />
          ))}

          <Xaxis
            className={styles.xAxis}
            transform={`translate(0, ${yScale.range()[0] + chartPadding / 2})`}
            data={data}
            xScale={xScale}
          />
        </Chart>

        <Legend color={colorFn.outflow} data={data.outflow} {...{ dataValue, dataLabel, dataKey }} />
      </div>
    );
  }

}


export default StackedChart;
