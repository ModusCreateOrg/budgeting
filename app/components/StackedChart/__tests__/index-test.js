import React from 'react';
import renderer from 'react-test-renderer';
import StackedChart from '../';

jest.mock('components/Legend', () => 'div');
jest.mock('components/Chart', () => 'div');
jest.mock('../Bar', () => 'div');
jest.mock('../Xaxis', () => 'div');

it('renders correctly', () => {
  const mockData = { inflow: [], outflow: [] };
  const mockTotals = { inflow: 1, outflow: 1 };

  const tree = renderer
    .create(<StackedChart dataLabel="test" dataKey="test" data={mockData} totals={mockTotals} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
