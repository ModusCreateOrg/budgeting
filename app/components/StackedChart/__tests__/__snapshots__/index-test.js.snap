// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="stackedChart"
>
  <div
    height={520}
    padding={10}
    transform="translate(10,10)"
    width={320}
  >
    <div
      colorFn={[Function]}
      data={Array []}
      transform="translate(0, 0)"
      width={128}
      yScale={[Function]}
    />
    <div
      colorFn={[Function]}
      data={Array []}
      transform="translate(151, 0)"
      width={128}
      yScale={[Function]}
    />
    <div
      data={
        Object {
          "inflow": Array [],
          "outflow": Array [],
        }
      }
      totals={
        Object {
          "inflow": 1,
          "outflow": 1,
        }
      }
      transform="translate(0, 443.3333333333333)"
      xScale={[Function]}
    />
  </div>
  <div
    color={[Function]}
    data={Array []}
    dataKey="test"
    dataLabel="test"
    dataValue="value"
    reverse={true}
  />
</div>
`;
